File: todolist/test_api_comparison.js
const fetch = require('node-fetch');
const fs = require('fs');
const path = require('path');

// 配置信息
const PORT1 = 3000; // Node.js 项目端口
const PORT2 = 3001; // Python 项目端口
const BASE_URL1 = `http://localhost:${PORT1}`; // Node.js 项目地址
const BASE_URL2 = `http://localhost:${PORT2}`; // Python 项目地址

// 测试结果输出文件
const TEST_RESULT_FILE = path.join(__dirname, 'api_comparison_result.json');
const TEST_REPORT_FILE = path.join(__dirname, 'api_comparison_report.md');

// 需要测试的接口列表
// 只测试查询类型的接口，忽略修改本地数据的接口
const TEST_INTERFACES = [
  // 节日相关接口
  {
    name: '获取节日配置',
    path: '/api/festival/config',
    method: 'GET',
    body: null
  },
  // 文件相关接口
  {
    name: '扫描文件',
    path: '/api/file/scan',
    method: 'GET',
    body: null
  },
  {
    name: '读取默认文件',
    path: '/api/file/read/todo.test.txt',
    method: 'GET',
    body: null
  },
  // 节假日相关接口
  {
    name: '获取节假日缓存',
    path: '/api/holiday/cache',
    method: 'GET',
    body: null
  },
  // 状态相关接口
  {
    name: '检查服务状态',
    path: '/api/check-status',
    method: 'GET',
    body: null
  },
  // 天气相关接口
  {
    name: '获取IP位置信息',
    path: '/api/weather/ip-location?forceRefresh=true',
    method: 'GET',
    body: null
  },
  {
    name: '获取天气区域编码',
    path: '/api/weather/weather-area-codes',
    method: 'GET',
    body: null
  },
  {
    name: '获取天气数据',
    path: '/api/weather/weather-info?weatherCode=101010100',
    method: 'GET',
    body: null
  }
];

/**
 * 发起HTTP请求
 * @param {string} url - 请求URL
 * @param {string} method - 请求方法
 * @param {Object} body - 请求体
 * @returns {Promise<Object>} 响应数据
 */
async function makeRequest(url, method, body) {
  try {
    const options = {
      method,
      headers: {
        'Content-Type': 'application/json'
      }
    };

    if (body) {
      options.body = JSON.stringify(body);
    }

    const response = await fetch(url, options);
    const data = await response.json();
    return {
      success: true,
      data,
      status: response.status
    };
  } catch (error) {
    return {
      success: false,
      error: error.message
    };
  }
}

/**
 * 过滤掉时间戳字段
 * @param {Object} obj - 要过滤的对象
 * @returns {Object} 过滤后的对象
 */
function filterTimestampFields(obj) {
  if (obj === null || typeof obj !== 'object') {
    return obj;
  }
  
  // 时间戳字段列表
  const timestampFields = ['timestamp', 'time', 'updatedAt', 'createdAt', 'expireAt', 'mtime', 'ttl'];
  
  const filteredObj = {};
  for (const key in obj) {
    if (obj.hasOwnProperty(key)) {
      // 忽略时间戳字段
      if (timestampFields.includes(key)) {
        continue;
      }
      // 递归过滤嵌套对象
      if (typeof obj[key] === 'object' && obj[key] !== null) {
        filteredObj[key] = filterTimestampFields(obj[key]);
      } else {
        filteredObj[key] = obj[key];
      }
    }
  }
  return filteredObj;
}

/**
 * 对比两个对象的差异
 * @param {Object} obj1 - 第一个对象
 * @param {Object} obj2 - 第二个对象
 * @returns {Object} 差异对象
 */
function compareObjects(obj1, obj2) {
  // 过滤时间戳字段
  const filteredObj1 = filterTimestampFields(obj1);
  const filteredObj2 = filterTimestampFields(obj2);
  
  const differences = {};

  // 检查obj1中的属性
  if (filteredObj1 && typeof filteredObj1 === 'object') {
    for (const key in filteredObj1) {
      if (filteredObj1.hasOwnProperty(key)) {
        if (!filteredObj2 || !filteredObj2.hasOwnProperty(key)) {
          differences[key] = {
            port1: filteredObj1[key],
            port2: undefined
          };
        } else if (typeof filteredObj1[key] === 'object' && typeof filteredObj2[key] === 'object' && filteredObj1[key] !== null && filteredObj2[key] !== null) {
          const nestedDiff = compareObjects(filteredObj1[key], filteredObj2[key]);
          if (Object.keys(nestedDiff).length > 0) {
            differences[key] = nestedDiff;
          }
        } else if (filteredObj1[key] !== filteredObj2[key]) {
          differences[key] = {
            port1: filteredObj1[key],
            port2: filteredObj2[key]
          };
        }
      }
    }
  }

  // 检查obj2中独有的属性
  if (filteredObj2 && typeof filteredObj2 === 'object') {
    for (const key in filteredObj2) {
      if (filteredObj2.hasOwnProperty(key) && (!filteredObj1 || !filteredObj1.hasOwnProperty(key))) {
        differences[key] = {
          port1: undefined,
          port2: filteredObj2[key]
        };
      }
    }
  }

  return differences;
}

/**
 * 运行单个接口测试
 * @param {Object} testInterface - 测试接口配置
 * @returns {Promise<Object>} 测试结果
 */
async function runInterfaceTest(testInterface) {
  console.log(`测试接口: ${testInterface.name}`);
  
  // 并行从两个端口获取数据
  const [result1, result2] = await Promise.all([
    makeRequest(
      `${BASE_URL1}${testInterface.path}`,
      testInterface.method,
      testInterface.body
    ),
    makeRequest(
      `${BASE_URL2}${testInterface.path}`,
      testInterface.method,
      testInterface.body
    )
  ]);

  // 对比结果
  let differences = {};
  if (result1.success && result2.success) {
    differences = compareObjects(result1.data, result2.data);
  } else if (result1.success !== result2.success) {
    // 一个端口请求成功，另一个失败，标记为有差异
    differences = {
      requestStatus: {
        port1: result1.success ? 'success' : 'failed',
        port2: result2.success ? 'success' : 'failed'
      }
    };
  }

  return {
    interface: testInterface,
    port1: result1,
    port2: result2,
    hasDifferences: Object.keys(differences).length > 0,
    differences
  };
}

/**
 * 生成测试报告
 * @param {Array} testResults - 测试结果数组
 */
function generateTestReport(testResults) {
  let report = '# API 接口对比测试报告\n\n';
  report += `测试时间: ${new Date().toLocaleString()}\n`;
  report += `测试端口1: ${PORT1}\n`;
  report += `测试端口2: ${PORT2}\n\n`;

  let totalTests = testResults.length;
  let passedTests = testResults.filter(result => !result.hasDifferences).length;
  let failedTests = totalTests - passedTests;

  report += `## 测试概览\n`;
  report += `总计测试: ${totalTests}\n`;
  report += `无差异: ${passedTests}\n`;
  report += `有差异: ${failedTests}\n\n`;

  report += `## 详细测试结果\n\n`;

  testResults.forEach((result, index) => {
    report += `### ${index + 1}. ${result.interface.name}\n`;
    report += `接口路径: ${result.interface.path}\n`;
    report += `请求方法: ${result.interface.method}\n\n`;

    if (!result.port1.success) {
      report += `**端口${PORT1}请求失败**: ${result.port1.error}\n\n`;
    } else if (!result.port2.success) {
      report += `**端口${PORT2}请求失败**: ${result.port2.error}\n\n`;
    } else if (result.hasDifferences) {
      report += `**发现差异**:\n`;
      report += `\`\`\`json\n`;
      report += JSON.stringify(result.differences, null, 2);
      report += `\`\`\`\n\n`;
    } else {
      report += `**结果一致**: 两个端口的响应完全相同\n\n`;
    }
  });

  fs.writeFileSync(TEST_REPORT_FILE, report);
  console.log(`测试报告已生成: ${TEST_REPORT_FILE}`);
}

/**
 * 主测试函数
 */
async function runApiComparisonTest() {
  console.log('开始 API 接口对比测试...');
  console.log(`测试端口1: ${PORT1}`);
  console.log(`测试端口2: ${PORT2}`);
  console.log('------------------------');

  // 并行运行所有接口测试
  console.log('并行测试中...');
  const testResults = await Promise.all(
    TEST_INTERFACES.map(async (testInterface) => {
      console.log(`开始测试: ${testInterface.name}`);
      const result = await runInterfaceTest(testInterface);
      console.log(`测试完成: ${testInterface.name} ${result.hasDifferences ? '【有差异】' : '【无差异】'}`);
      console.log('------------------------');
      return result;
    })
  );

  // 保存测试结果
  const testResultData = {
    testTime: new Date().toISOString(),
    port1: PORT1,
    port2: PORT2,
    totalTests: testResults.length,
    passedTests: testResults.filter(result => !result.hasDifferences).length,
    failedTests: testResults.filter(result => result.hasDifferences).length,
    testResults
  };

  fs.writeFileSync(TEST_RESULT_FILE, JSON.stringify(testResultData, null, 2));
  console.log(`测试结果已保存: ${TEST_RESULT_FILE}`);

  // 生成测试报告
  generateTestReport(testResults);

  console.log('------------------------');
  console.log('API 接口对比测试完成!');
  console.log(`总计测试: ${testResults.length}`);
  console.log(`无差异: ${testResults.filter(result => !result.hasDifferences).length}`);
  console.log(`有差异: ${testResults.filter(result => result.hasDifferences).length}`);
  console.log(`详细报告请查看: ${TEST_REPORT_FILE}`);
}

// 运行测试
if (require.main === module) {
  runApiComparisonTest().catch(error => {
    console.error('测试过程中发生错误:', error);
  });
}

module.exports = {
  runApiComparisonTest,
  TEST_INTERFACES
};

File: todolist-python/test_migration/test_api_comparison.py
import http.client
import json
import os
import time
import argparse
from datetime import datetime

# 命令行参数解析
def parse_args():
    parser = argparse.ArgumentParser(description='API 接口对比测试脚本')
    parser.add_argument('--base-port', type=int, default=3001, help='基准项目端口')
    parser.add_argument('--migrate-port', type=int, default=3002, help='迁移项目端口')
    parser.add_argument('--base-name', type=str, default='Python', help='基准项目名称')
    parser.add_argument('--migrate-name', type=str, default='Tauri (Rust)', help='迁移项目名称')
    parser.add_argument('--output-dir', type=str, default=os.path.dirname(__file__), help='测试结果输出目录')
    return parser.parse_args()

# 解析命令行参数
args = parse_args()

# 配置信息
BASE_PORT = args.base_port  # 基准项目端口
MIGRATE_PORT = args.migrate_port  # 迁移项目端口
BASE_URL = f"localhost:{BASE_PORT}"  # 基准项目地址
MIGRATE_URL = f"localhost:{MIGRATE_PORT}"  # 迁移项目地址

# 项目名称
BASE_NAME = args.base_name
MIGRATE_NAME = args.migrate_name

# 测试结果输出文件
TEST_RESULT_FILE = os.path.join(args.output_dir, 'api_comparison_result.json')

# 需要测试的接口列表
# 只测试查询类型的接口，忽略修改本地数据的接口
TEST_INTERFACES = [
    # 节日相关接口
    {
        "name": "获取节日配置",
        "path": "/api/festival/config",
        "method": "GET",
        "body": None
    },
    # 文件相关接口
    {
        "name": "扫描文件",
        "path": "/api/file/scan",
        "method": "GET",
        "body": None
    },
    {
        "name": "读取默认文件",
        "path": "/api/file/read/todo.test.txt",
        "method": "GET",
        "body": None
    },
    # 节假日相关接口
    {
        "name": "获取节假日缓存",
        "path": "/api/holiday/cache",
        "method": "GET",
        "body": None
    },
    # 状态相关接口
    {
        "name": "检查服务状态",
        "path": "/api/check-status",
        "method": "GET",
        "body": None
    },
    # 天气相关接口
    {
        "name": "获取 IP 位置信息",
        "path": "/api/weather/ip-location?forceRefresh=true",
        "method": "GET",
        "body": None
    },
    {
        "name": "获取天气区域编码",
        "path": "/api/weather/weather-area-codes",
        "method": "GET",
        "body": None
    },
    {
        "name": "获取天气数据",
        "path": "/api/weather/weather-info?weatherCode=101010100",
        "method": "GET",
        "body": None
    }
]


class HTTPError(Exception):
    """HTTP错误异常"""
    pass


def make_request(host, path, method="GET", body=None):
    """
    发起 HTTP 请求
    :param host: 主机地址，格式为 "host:port"
    :param path: 请求路径，包含查询参数
    :param method: 请求方法
    :param body: 请求体
    :return: 响应数据字典
    """
    try:
        # 解析主机和端口
        host_port = host.split(":")
        if len(host_port) != 2:
            raise ValueError(f"Invalid host format: {host}")
        host_name = host_port[0]
        port = int(host_port[1])
        
        # 创建连接
        conn = http.client.HTTPConnection(host_name, port, timeout=15)
        
        # 准备请求头
        headers = {
            'Content-Type': 'application/json',
            'Accept': 'application/json; charset=utf-8'
        }
        
        # 发送请求
        if body:
            body_str = json.dumps(body)
            conn.request(method, path, body_str, headers)
        else:
            conn.request(method, path, headers=headers)
        
        # 获取响应
        response = conn.getresponse()
        status = response.status
        data = response.read().decode('utf-8')
        conn.close()
        
        # 解析响应数据
        try:
            json_data = json.loads(data)
            return {
                "success": True,
                "data": json_data,
                "status": status
            }
        except json.JSONDecodeError as e:
            return {
                "success": False,
                "error": f"JSON 解析失败：{str(e)}",
                "status": status
            }
    except Exception as e:
        return {
            "success": False,
            "error": str(e)
        }


def filter_timestamp_fields(obj):
    """
    过滤掉时间戳字段
    :param obj: 要过滤的对象
    :return: 过滤后的对象
    """
    if obj is None or not isinstance(obj, (dict, list)):
        return obj
    
    # 时间戳字段列表
    timestamp_fields = ['timestamp', 'time', 'updatedAt', 'createdAt', 'expireAt', 'mtime', 'ttl']
    
    if isinstance(obj, dict):
        filtered_obj = {}
        for key, value in obj.items():
            # 忽略时间戳字段
            if key in timestamp_fields:
                continue
            # 递归过滤嵌套对象
            if isinstance(value, (dict, list)):
                filtered_obj[key] = filter_timestamp_fields(value)
            else:
                filtered_obj[key] = value
        return filtered_obj
    elif isinstance(obj, list):
        return [filter_timestamp_fields(item) for item in obj]


def compare_objects(obj1, obj2):
    """
    对比两个对象的差异
    :param obj1: 第一个对象
    :param obj2: 第二个对象
    :return: 差异对象
    """
    # 过滤时间戳字段
    filtered_obj1 = filter_timestamp_fields(obj1)
    filtered_obj2 = filter_timestamp_fields(obj2)
    
    differences = {}
    
    # 检查 obj1 中的属性
    if filtered_obj1 and isinstance(filtered_obj1, dict):
        for key, value1 in filtered_obj1.items():
            if not filtered_obj2 or not isinstance(filtered_obj2, dict) or key not in filtered_obj2:
                differences[key] = {
                    "service1": value1,
                    "service2": None
                }
            elif isinstance(value1, (dict, list)) and isinstance(filtered_obj2[key], (dict, list)):
                nested_diff = compare_objects(value1, filtered_obj2[key])
                if nested_diff:
                    differences[key] = nested_diff
            elif value1 != filtered_obj2[key]:
                differences[key] = {
                    "service1": value1,
                    "service2": filtered_obj2[key]
                }
    
    # 检查 obj2 中独有的属性
    if filtered_obj2 and isinstance(filtered_obj2, dict):
        for key, value2 in filtered_obj2.items():
            if not filtered_obj1 or not isinstance(filtered_obj1, dict) or key not in filtered_obj1:
                differences[key] = {
                    "service1": None,
                    "service2": value2
                }
    
    return differences


async def run_interface_test(test_interface):
    """
    运行单个接口测试
    :param test_interface: 测试接口配置
    :return: 测试结果
    """
    print(f"测试接口：{test_interface['name']}")
    
    # 从两个服务获取数据
    base_result = make_request(BASE_URL, test_interface['path'], test_interface['method'], test_interface['body'])
    migrate_result = make_request(MIGRATE_URL, test_interface['path'], test_interface['method'], test_interface['body'])
    
    # 对比结果
    differences = {}
    if base_result['success'] and migrate_result['success']:
        differences = compare_objects(base_result['data'], migrate_result['data'])
    elif base_result['success'] != migrate_result['success']:
        # 一个服务请求成功，另一个失败，标记为有差异
        differences = {
            "requestStatus": {
                "baseService": "success" if base_result['success'] else "failed",
                "migrateService": "success" if migrate_result['success'] else "failed"
            }
        }
    
    return {
        "interface": test_interface,
        "baseService": base_result,
        "migrateService": migrate_result,
        "hasDifferences": bool(differences),
        "differences": differences
    }


def generate_test_report(test_results):
    """
    生成测试报告（仅控制台输出）
    :param test_results: 测试结果数组
    """
    total_tests = len(test_results)
    passed_tests = len([r for r in test_results if not r['hasDifferences']])
    failed_tests = total_tests - passed_tests
    
    print('\n=== 测试概览 ===')
    print(f'总计测试：{total_tests}')
    print(f'无差异：{passed_tests}')
    print(f'有差异：{failed_tests}')
    print('\n=== 详细测试结果 ===')
    
    for i, result in enumerate(test_results, 1):
        print(f'\n{i}. {result["interface"]["name"]}')
        print(f'接口路径：{result["interface"]["path"]}')
        print(f'请求方法：{result["interface"]["method"]}')
        
        if not result["baseService"]["success"]:
            print(f'**{BASE_NAME} 请求失败**: {result["baseService"]["error"]}')
        elif not result["migrateService"]["success"]:
            print(f'**{MIGRATE_NAME} 请求失败**: {result["migrateService"]["error"]}')
        elif result["hasDifferences"]:
            print('**发现差异**:')
            print(json.dumps(result["differences"], ensure_ascii=False, indent=2))
        else:
            print('**结果一致**: 两个服务的响应完全相同')


async def run_api_comparison_test():
    """
    主测试函数
    """
    print('开始 API 接口对比测试...')
    print(f'基准项目: {BASE_NAME} (端口:{BASE_PORT})')
    print(f'迁移项目: {MIGRATE_NAME} (端口:{MIGRATE_PORT})')
    print('------------------------')
    
    # 运行所有接口测试
    print('测试中...')
    test_results = []
    for test_interface in TEST_INTERFACES:
        print(f'开始测试：{test_interface["name"]}')
        result = await run_interface_test(test_interface)
        print(f'测试完成：{test_interface["name"]} {"【有差异】" if result["hasDifferences"] else "【无差异】"}')
        print('------------------------')
        test_results.append(result)
    
    # 保存测试结果
    test_result_data = {
        "testTime": datetime.now().isoformat(),
        "baseService": BASE_NAME,
        "migrateService": MIGRATE_NAME,
        "basePort": BASE_PORT,
        "migratePort": MIGRATE_PORT,
        "totalTests": len(test_results),
        "passedTests": len([r for r in test_results if not r['hasDifferences']]),
        "failedTests": len([r for r in test_results if r['hasDifferences']]),
        "testResults": test_results
    }
    
    with open(TEST_RESULT_FILE, 'w', encoding='utf-8') as f:
        json.dump(test_result_data, f, ensure_ascii=False, indent=2)
    print(f'测试结果已保存：{TEST_RESULT_FILE}')
    
    # 生成测试报告
    generate_test_report(test_results)
    
    print('------------------------')
    print('API 接口对比测试完成!')
    print(f'总计测试：{len(test_results)}')
    print(f'无差异：{len([r for r in test_results if not r["hasDifferences"]])}')
    print(f'有差异：{len([r for r in test_results if r["hasDifferences"]])}')
    print(f'测试结果已保存：{TEST_RESULT_FILE}')


if __name__ == '__main__':
    import asyncio
    asyncio.run(run_api_comparison_test())
